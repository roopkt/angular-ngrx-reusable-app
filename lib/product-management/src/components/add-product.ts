import { Component, Input, ElementRef, EventEmitter, ChangeDetectionStrategy } from '@angular/core';
import { Output } from '@angular/core';
import { ViewChild } from '@angular/core';

@Component({
    selector: 'pm-add-product',
    template: `
        <div>
        <form (submit)="addProductClick($event)">
            <p>Product Name:<input type="text" #prodName name="prodName" id="prodName"></p>
        </form>
        </div>
    `,
    changeDetection: ChangeDetectionStrategy.OnPush,
})
export class AddProductComponent {
    @Output() addProduct = new EventEmitter();
    @ViewChild('prodName') prodName: ElementRef;
    constructor() {
        
    }
    addProductClick(event: KeyboardEvent) {
        event.stopPropagation();
        event.preventDefault();
        const name = this.prodName.nativeElement.value;
        this.addProduct.emit(name);
    }
}

