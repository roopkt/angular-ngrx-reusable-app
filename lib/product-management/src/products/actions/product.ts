import { Action } from '@ngrx/store';
import { Product } from '../models/product';


export const ADD = '[Releases] Add';

export class Add implements Action {
    readonly type = ADD;
    constructor(public payload: Product) { }
}

export type Actions = Add;



