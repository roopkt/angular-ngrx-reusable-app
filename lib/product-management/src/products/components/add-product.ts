import { Component, Input, ElementRef, EventEmitter, ChangeDetectionStrategy } from '@angular/core';
import { Output } from '@angular/core';

@Component({
    selector: 'pm-add-product',
    template: `
        <div>
        <form (ngSubmit)="addProductClick($event)">
            <p>Product Name:<input type="text" name="prodName" id="prodName"></p>
        </form>
        </div>
    `,
    changeDetection: ChangeDetectionStrategy.OnPush,
})
export class AddProductComponent {
    @Output() addProduct = new EventEmitter();

    constructor(public el: ElementRef) {

    }

    addProductClick(event) {
        const name = this.el.nativeElement.value;
        this.addProduct.emit(name);
    }
}

