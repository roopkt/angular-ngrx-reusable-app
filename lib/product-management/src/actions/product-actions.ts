import { Action } from '@ngrx/store';
import { Product } from '../models/product';

export enum ProductActionTypes {
    ADD = '[Products] Add',
    LOAD = '[Products] Load',
}

export class Add implements Action {
    readonly type = ProductActionTypes.ADD;
    constructor(public payload: Product) { }
}

export class Load implements Action {
    readonly type = ProductActionTypes.LOAD;
    constructor() { }
}

export type ProductActions = Add | Load;



