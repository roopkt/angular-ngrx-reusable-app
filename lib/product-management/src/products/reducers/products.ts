import { createEntityAdapter, EntityAdapter, EntityState } from '@ngrx/entity';
import * as productActions from '../actions/Product';
import { Product } from '../models/product';

export interface State extends EntityState<Product> {
    selectedProductId: string | null;
}

/**
 * createEntityAdapter creates many an object of helper
 * functions for single or multiple operations
 * against the dictionary of records. The configuration
 * object takes a record id selector function and
 * a sortComparer option which is set to a compare
 * function if the records are to be sorted.
 */
export const adapter: EntityAdapter<Product> = createEntityAdapter<Product>({
    selectId: (product: Product) => product.id,
    sortComparer: false,
});

/** getInitialState returns the default initial state
 * for the generated entity state. Initial state
 * additional properties can also be defined.
*/
export const initialState: State = adapter.getInitialState({
    selectedProductId: null,
});

export function reducer(state = initialState,
    action: productActions.Actions): State {
    switch (action.type) {
        case productActions.ADD: {
            return {
                ...adapter.addOne(action.payload, state),
                selectedProductId: state.selectedProductId
            };
        }
        default: {
            return state;
        }
    }
}

export const getSelectedId = (state: State) => state.selectedProductId;
