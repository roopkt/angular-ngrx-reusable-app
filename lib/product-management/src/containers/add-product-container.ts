import { ChangeDetectionStrategy, Component, OnInit } from '@angular/core';
import { Store, select } from '@ngrx/store';
import { Product } from '../models/product';
import { Observable } from 'rxjs/Observable';

import { Add, Load } from '../actions/product-actions';
import { getAllProducts, State } from '../reducers';

@Component({
    selector: 'pm-add-product-container',
    exportAs: 'pmcAddProduct',
    template: `
        <div>
        <pm-add-product (addProduct)="addProduct($event)"></pm-add-product>
        </div>
    `,
    changeDetection: ChangeDetectionStrategy.OnPush,
})
export class AddProductContainerComponent implements OnInit {
    products$: Observable<Product[]>;
    constructor(private store: Store<State>) {
        this.products$ = store.pipe(select(getAllProducts));
    }

    ngOnInit(): void {
        this.store.dispatch(new Load());
    }

    addProduct(name: string): void {
        this.store.dispatch(new Add({
            id: (Math.random()).toString().substr(4),
            name: name
        }));
    }
}
