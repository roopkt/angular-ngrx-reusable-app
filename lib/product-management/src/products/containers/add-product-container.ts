import { Store } from '@ngrx/store';
import { State } from '../reducers';
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { Add } from '../actions/Product';

@Component({
    selector: 'pm-add-product-container',
    template: `
        <div>
        <pm-add-product (addProduct)="addProduct($event)"></pm-add-product>
        </div>
    `,
    changeDetection: ChangeDetectionStrategy.OnPush,
})
export class AddProductContainerComponent {
    constructor(private store: Store<State>) {
    }

    addProduct(name: string) {
        this.store.dispatch(new Add({
            id: Math.random().toString(),
            name: name
        }));
    }
}
