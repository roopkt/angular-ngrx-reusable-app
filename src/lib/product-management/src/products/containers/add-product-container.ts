import { Store } from '@ngrx/store';
import {State} from '../reducers';
import {Add} from '../actions/product';
import { ChangeDetectionStrategy, Component } from '@angular/core';

@Component({
    selector: 'rm-release-detail',
    template: `
        <div>
        <pm-release-detail (addProduct)="addProduct($event)"></pm-release-detail>
        </div>
    `,
    changeDetection: ChangeDetectionStrategy.OnPush,
})
export class AddProductContainerComponent {
    constructor(private store: Store<State>) {
    }

    addProduct(name: string) {
        this.store.dispatch(new Add({
            id: Math.random().toString(),
            name: name
        }));
    }
}
