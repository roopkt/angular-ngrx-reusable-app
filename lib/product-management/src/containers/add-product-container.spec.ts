import { ComponentFixture, TestBed } from '@angular/core/testing';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import { RouterTestingModule } from '@angular/router/testing';
import { combineReducers, Store, StoreModule } from '@ngrx/store';

import { Load } from '../actions/product-actions';
import * as fromProducts from '../reducers';
import { AddProductComponent } from './../components/add-product';
import { AddProductContainerComponent } from './add-product-container';

describe('Add Product Page', () => {
  let fixture: ComponentFixture<AddProductContainerComponent>;
  let store: Store<fromProducts.State>;
  let instance: AddProductContainerComponent;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [
        NoopAnimationsModule,
        StoreModule.forRoot({
          books: combineReducers(fromProducts.reducers),
        }),
        RouterTestingModule,
      ],
      declarations: [
        AddProductContainerComponent,
        AddProductComponent,
      ],
    });

    fixture = TestBed.createComponent(AddProductContainerComponent);
    instance = fixture.componentInstance;
    store = TestBed.get(Store);

    spyOn(store, 'dispatch').and.callThrough();
  });

  it('should compile', () => {
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should dispatch a productActions.Load on init', () => {
    const action = new Load();

    fixture.detectChanges();

    expect(store.dispatch).toHaveBeenCalledWith(action);
  });
});
