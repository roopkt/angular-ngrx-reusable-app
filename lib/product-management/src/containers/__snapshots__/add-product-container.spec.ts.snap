// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add Product Page should compile 1`] = `
<pm-add-product-container
  products$={[Function Store]}
  store={[Function Store]}
>
  
          <div>
    
          
    <pm-add-product>
      
          
      <div>
        
          
        <form>
          
              
          <p>
            Product Name:
            <input
              id="prodName"
              name="prodName"
              type="text"
            />
          </p>
          
          
        </form>
        
          
      </div>
      
      
    </pm-add-product>
    
          
  </div>
      
</pm-add-product-container>
`;
